File: Collections/HeapOperations.vb
Public Class HeapOperations(Of T)

    Public Sub New(ByVal Comparer As IComparer(Of T))
        Me.Comparer = Comparer
    End Sub

    Public Sub New()
        Me.New(Generic.Comparer(Of T).Default)
    End Sub

    Public Sub Heapify(ByVal List As IList(Of T), ByVal Count As Integer, ByVal Index As Integer)
        Do
            Dim MinIndex = LeftChildren(Index)
            If MinIndex >= Count Then
                Exit Do
            End If

            Dim Min = List.Item(MinIndex)
            Dim I = RightChildren(Index)
            Dim T As T
            If I < Count Then
                T = List.Item(I)
                If Me.Comparer.Compare(T, Min) < 0 Then
                    MinIndex = I
                    Min = T
                End If
            End If

            T = List.Item(Index)
            If Me.Comparer.Compare(Min, T) < 0 Then
                List.Item(Index) = Min
                List.Item(MinIndex) = T

                Index = MinIndex
            Else
                Exit Do
            End If
        Loop
    End Sub

    Public Sub MakeHeap(ByVal List As IList(Of T), ByVal Count As Integer)
        For I As Integer = Parent(Count - 1) To 0 Step -1
            Me.Heapify(List, Count, I)
        Next
    End Sub

    Public Sub BubbleUp(ByVal List As IList(Of T), ByVal Count As Integer, ByVal Index As Integer)
        Do
            If Index = 0 Then
                Exit Do
            End If

            Dim T = List.Item(Index)
            Dim PI = Parent(Index)
            Dim P = List.Item(PI)

            If Me.Comparer.Compare(P, T) > 0 Then
                List.Item(PI) = T
                List.Item(Index) = P
                Index = PI
            Else
                Exit Do
            End If
        Loop
    End Sub

    Public Shared Function Parent(ByVal Index As Integer) As Integer
        Return (Index - 1) \ 2
    End Function

    Public Shared Function LeftChildren(ByVal Index As Integer) As Integer
        Return 2 * Index + 1
    End Function

    Public Shared Function RightChildren(ByVal Index As Integer) As Integer
        Return 2 * Index + 2
    End Function

    Private ReadOnly Comparer As IComparer(Of T)

End Class

File: Collections/HeapOperations.cs
using System.Collections.Generic;

namespace Ks.Common
{
    public class HeapOperations<T>
    {
        public HeapOperations(IComparer<T> Comparer)
        {
            this.Comparer = Comparer;
        }

        public HeapOperations() : this(System.Collections.Generic.Comparer<T>.Default)
        {
        }

        public void Heapify(IList<T> List, int Count, int Index)
        {
            while (true)
            {
                var MinIndex = LeftChild(Index);
                if (MinIndex >= Count)
                {
                    break;
                }

                var Min = List[MinIndex];
                var I = RightChild(Index);
                var T = default(T);
                if (I < Count)
                {
                    T = List[I];
                    if (this.Comparer.Compare(T, Min) < 0)
                    {
                        MinIndex = I;
                        Min = T;
                    }
                }

                T = List[Index];
                if (this.Comparer.Compare(Min, T) < 0)
                {
                    List[Index] = Min;
                    List[MinIndex] = T;
                    Index = MinIndex;
                }
                else
                {
                    break;
                }
            }
        }

        public void MakeHeap(IList<T> List, int Count)
        {
            for (var I = Parent(Count - 1); I >= 0; I--)
            {
                this.Heapify(List, Count, I);
            }
        }

        public void BubbleUp(IList<T> List, int Count, int Index)
        {
            while (true)
            {
                if (Index == 0)
                {
                    break;
                }

                var T = List[Index];
                var PI = Parent(Index);
                var P = List[PI];
                if (this.Comparer.Compare(P, T) > 0)
                {
                    List[PI] = T;
                    List[Index] = P;
                    Index = PI;
                }
                else
                {
                    break;
                }
            }
        }

        public static int Parent(int Index)
        {
            return (Index - 1) / 2;
        }

        public static int LeftChild(int Index)
        {
            return (2 * Index) + 1;
        }

        public static int RightChild(int Index)
        {
            return (2 * Index) + 2;
        }

        private readonly IComparer<T> Comparer;
    }
}
